### ISSUE TITLE
Document encrypt.* values need to go in bootstrap.{yml|properites}

### ISSUE THREAD

**Issue by bplies-IAS** [spring-cloud/spring-cloud-config#895]
Split from #865 as a separate issue.

While attempting an upgrade of spring-cloud-config-server from 1.2.2.RELEASE to 1.3.2.RELEASE it stopped decrypting properties when calling the basic
$ curl http://user:password@localhost:PORT/APP-NAME/PROFILE/LABEL

Using the same JKS store and everything.

I've done some debugging and have been able to reproduce the problem and narrow down what I believe is the cause.

Spring Boot:  1.5.9
Spring Cloud Config Server: 1.3.3.RELEASE 
Cloud Dependencies:  Dalston.SR4
(Matches **Release Trains** described [here](http://projects.spring.io/spring-cloud/))

When using a JKS to encrypt/decrypt, upon startup spring-cloud-config-server-1.3.3.RELEASE-sources.jar!/org/springframework/cloud/config/server/config/EncryptionAutoConfiguration.java
provides configuration as follows:

```
	@Configuration
	@ConditionalOnProperty(value = "spring.cloud.config.server.encrypt.enabled", matchIfMissing = true)
	protected static class EncryptorConfiguration {

		@Autowired(required = false)
		private TextEncryptorLocator locator;

		@Autowired
		private TextEncryptor encryptor;

		@Bean
		@ConditionalOnMissingBean
		public EnvironmentEncryptor environmentEncryptor() {
			TextEncryptorLocator locator = this.locator;
			if (locator == null) {
				locator = new SingleTextEncryptorLocator(encryptor);
			}
			return new CipherEnvironmentEncryptor(locator);
		}

	}

	@Configuration
	@ConditionalOnClass(RsaSecretEncryptor.class)
	@ConditionalOnProperty(prefix = "encrypt.keyStore", value = "location", matchIfMissing = false)
	protected static class KeyStoreConfiguration {

		@Autowired
		private KeyProperties key;

		@Bean
		@ConditionalOnMissingBean
		public TextEncryptorLocator textEncryptorLocator() {
			KeyStore keyStore = this.key.getKeyStore();
			KeyStoreTextEncryptorLocator locator = new KeyStoreTextEncryptorLocator(
					new KeyStoreKeyFactory(keyStore.getLocation(), keyStore.getPassword().toCharArray()),
					keyStore.getSecret(), keyStore.getAlias());
			locator.setRsaAlgorithm(this.key.getRsa().getAlgorithm());
			locator.setSalt(this.key.getRsa().getSalt());
			locator.setStrong(this.key.getRsa().isStrong());
			return locator;
		}

	}
```

The problem is that `EncryptorConfiguration` evaluated first `spring.cloud.config.server.encrypt.enabled=true` and because the `TextEncryptorLocator` was NULL set it as `SingleTextEncryptorLocator `and used that to construct `CipherEnvironmentEncryptor`.

Afterwards, `KeyStoreConfiguration` evaluates because of `@ConditionalOnProperty(prefix = "encrypt.keyStore", value = "location", matchIfMissing = false)` and attempts to create a `KeyStoreTextEncryptorLocator` because 
```
@ConditionalOnMissingBean
public TextEncryptorLocator textEncryptorLocator()
```
however the `CipherEnvironmentEncryptor` is already created without access to `KeyStoreTextEncryptorLocator` the and so ends up unable to decrypt with the JKS.

Setting breakpoints on:
- CipherEnvironmentEncryptor.java#50
- EnvironmentController.java#107
- EncryptionAutoConfiguration.java#93

Starting the embedded Config Server application in debug mode can confirm how things were wired together.  Then starting another App that consults the Config Server and hitting the breakpoint you can confirm that the encrypted properties will not be decrypted.

I'm guessing there just is not an automated test covering JKS with decryption.  There is `KeyStoreTextEncryptorLocatorTests` but I don't see that `KeyStoreTextEncryptorLocator` ever successfully gets wired and used by `CipherEnvironmentEncryptor`.  `CipherEnvironmentEncryptorTests` only exercises `SingleTextEncryptorLocator`

Retried Dalston.SR5 (s-c-config 1.3.4) with JKS and it still isn't decrypting.

**Comment by spencergibb** [comment]
The main thing I asked for if you opened an issue was for a minimal sample that recreates the problem, since I am unable to do so.

**Comment by bplies-IAS** [comment]
I made a demo https://github.com/bplies-IAS/spring-cloud-config-895  check the README for instructions.

**Comment by spencergibb** [comment]
`encrypt.*` needs to be in bootstrap.properties.

See #887

**Comment by bplies-IAS** [comment]
Unfortunately I have not had success with `spring.cloud.config.server.encrypt.enabled` true or false in either `bootstrap.yml` or `bootstrap.properties`

When true the result is the same.  When false the payload is still encrypted but also still contains the `{cipher}{key:mytestkey}` prefix.

**Comment by spencergibb** [comment]
@bplies-IAS I don't understand what the new problem is. You're referenced project worked correctly when I moved `encrypt.*` to bootstrap.

**Comment by bplies-IAS** [comment]
@spencergibb Still a mystery.  I've tried `./gradlew --refresh-dependencies clean bootRun` to clear the workspace to be extra sure there isn't some residual gunk.  Didn't help.  I'll push what I currently have to the demo showing the explicit property and enabling several Actuator endpoints.  Maybe I typo'd something?  Property path not exactly correct?  

Otherwise I'm willing to have a screen share session where we could compare screens.  Short of that I don't know what else we can do other than compare hashes of various jars.  Does `./gradlew dependencies --configuration runtime` look correct to you?  

Given the lack of other corroborating reports of something this fundamental I'm inclined to believe this would likely be something local. My `env` doesn't show any env vars that would influence Spring Boot.  I don't have any other instance already occupying the port.

```
Darwin Kernel Version 15.6.0
java version "1.8.0_73"
JCE unlimited strength encryption
```

**Comment by bplies-IAS** [comment]
Just following up on this.  I do not believe this is a mere documentation issue as the configuration properties have no effect and debugging can in fact show a different behavior.  My position remains that `EncryptionAutoConfiguration` has a degree of non-deterministic behavior of Bean construction sequence among environments as demonstrated between @spencergibb and myself.  It will work fine for some, will not work for others.  Two `@Configuration` elements are both simultaneously eligible to evaluate by their conditionals and the sequence matters.

**Comment by spencergibb** [comment]
You did not move 
```
encrypt:
  keyStore:
    location: classpath:server.jks
    password: letmein
#    alias: mykey
    secret: changeme
```
https://github.com/bplies-IAS/spring-cloud-config-895/blob/88e7332e59249fef38e0480d9048244efa421021/src/main/resources/application.yml#L16-L21

to bootstrap.yml

**Comment by bplies-IAS** [comment]
I understand now.  Yes you are and have been correct.  The entire block, and not nested under `spring.cloud.config` prefix, is necessary and must be located in `bootstrap.yml`.   Thank you for clarifying

**Comment by ryanjbaxter** [comment]
Fixed via 484b74b
